### PR TITLE
[mono][HybridGlobalization] Fix ShortDatePattern year format to be "yyyy"

### LINKED ISSUE #99515: [iOS/HybridGlobalization] Force short date pattern to use yyyy
### Description

When using `HybridGlobalization` on iOS, the `shortdate` format does only contain just either one or two `y`'s, which result in only a two-digit year when formatting dates.

### Reproduction Steps

Set `<HybridGlobalization>true</HybridGlobalization>` in iOS csproj file and format some date with "short date" format (e.g. `DateTime.Now.ToString("d")`)

### Expected behavior

new CultureInfo("de-DE").DateTimeFormat.ShortDatePattern == "dd.MM.yyyy"
new CultureInfo("en-US").DateTimeFormat.ShortDatePattern == "M/d/yyyy"
new CultureInfo("fa-IR").DateTimeFormat.ShortDatePattern == "yyyy/M/d"

### Actual behavior

new CultureInfo("de-DE").DateTimeFormat.ShortDatePattern == "dd.MM.yy"
new CultureInfo("en-US").DateTimeFormat.ShortDatePattern == "M/d/yy"
new CultureInfo("fa-IR").DateTimeFormat.ShortDatePattern == "y/M/d"

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 8

### Other information

A similar problem was reported a while back for Linux [here](https://github.com/dotnet/runtime/issues/26088) and fixed [here](https://github.com/dotnet/coreclr/pull/18316). To code is still in the repo: https://github.com/dotnet/runtime/blob/3eb8c7f1086b79b28a27b57a935f97be3b7fcccb/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Icu.cs#L166-L251

### PR DESCRIPTION
**Pull request by matouskozak**
The globalization `DateTimeFormat.ShortDatePattern` API is returning short year format instead of "yyyy" format when Hybrid Globalization is enabled on iOS (previously discussed for Linux in https://github.com/dotnet/runtime/issues/26088). This PR overwrites the short year format ("y"/"yy") with "yyyy".

Fixes https://github.com/dotnet/runtime/issues/99515. Needs backport to .NET 8.

---

Contributed: @mkhamoyan

### CODE REVIEW

**Comment by tipa** [comment]
Thanks for fixing this so quickly. Can it be ruled out that the iOS native APIs return "yyyy" as the `shortFormatString`? Because if the native API returns "yyyy", I think [the added code](https://github.com/dotnet/runtime/pull/99908/commits/e2ac11214ec22de16b6efc0476dce91b5d909281) would replace it with "yyyyyyyy"

**Comment by matouskozak** [comment]
> Thanks for fixing this so quickly. Can it be ruled out that the iOS native APIs return "yyyy" as the `shortFormatString`? Because if the native API returns "yyyy", I think [the added code](https://github.com/dotnet/runtime/pull/99908/commits/e2ac11214ec22de16b6efc0476dce91b5d909281) would replace it with "yyyyyyyy"

Yes, you are right. It seems that the iOS native APIs do return only "y"/"yy" for the tested scenarios but I will push a change so that we can be safe.

**Comment by akoeplinger** [comment]
I haven't looked too closely but why is `FixDefaultShortDatePattern` in managed not kicking in? I'd prefer reusing that instead of dealing with it in native code.

**Comment by mkhamoyan** [comment]
> I haven't looked too closely but why is `FixDefaultShortDatePattern` in managed not kicking in? I'd prefer reusing that instead of dealing with it in native code.

Currently `FixDefaultShortDatePattern` is not called for hybrid mode https://github.com/dotnet/runtime/blob/589beb0b2af3edd1b272f8ac21ca5e3fb142bdd0/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs#L16-L19
We can reuse `FixDefaultShortDatePattern`, but we will need to modify it as apple native api now returns "y" or "yy".

**Comment by matouskozak** [comment]
> > I haven't looked too closely but why is `FixDefaultShortDatePattern` in managed not kicking in? I'd prefer reusing that instead of dealing with it in native code.
> 
> Currently `FixDefaultShortDatePattern` is not called for hybrid mode
> 
> https://github.com/dotnet/runtime/blob/589beb0b2af3edd1b272f8ac21ca5e3fb142bdd0/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs#L16-L19
> 
> 
> We can reuse `FixDefaultShortDatePattern`, but we will need to modify it as apple native api now returns "y" or "yy".

Yes, that is right. I was thinking about tinkering the `FixDefaultShortDatePattern` so that it works for "y" case but I wasn't sure if it is the best approach as it is used outside Hybrid as well and we will need backport to .NET 8.

Let me know if you would prefer this approach and I can give it a go.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.iOS.cs [hunk 1]
```diff
@@ -18,6 +18,15 @@ private bool LoadCalendarDataFromNative(string localeName, CalendarId calendarId
             sNativeName = GetCalendarInfoNative(localeName, calendarId, CalendarDataType.NativeName);
             sMonthDay = GetCalendarInfoNative(localeName, calendarId, CalendarDataType.MonthDay);
             saShortDates = GetCalendarInfoNative(localeName, calendarId, CalendarDataType.ShortDates).Split("||");
+            // Handle ShortDatePattern to have "yyyy" year format
+            List<string> shortDatePatternList = new List<string>(saShortDates);
+            for (int i = 0; i < shortDatePatternList.Count; i++)
+            {
+                shortDatePatternList[i] = NormalizeDatePattern(shortDatePatternList[i]);
```
Comment (ilonatommy) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.iOS.cs:25]: If we skiped `NormalizeDatePattern` then wouldn't it work? Or is it that it `NormalizeDatePattern` handles only `y -> yyyy` and `FixDefaultShortDatePattern` only `yy -> yyyy`?

File: src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs [hunk 2]
```diff
@@ -11,9 +11,9 @@ public class DateTimeFormatInfoShortDatePattern
         public static IEnumerable<object[]> ShortDatePattern_Get_TestData_HybridGlobalization()
         {
             // see the comments on the right to check the non-Hybrid result, if it differs
-            yield return new object[] { "ar-SA", "d\u200f/M\u200f/yyyy" }; // "d\u200f/M\u200f/yyyy g"
+            yield return new object[] { "ar-SA", PlatformDetection.IsHybridGlobalizationOnBrowser ? "d\u200f/M\u200f/yyyy" : "d MMM، yyyy g" }; // "d\u200f/M\u200f/yyyy g"
```
Comment (ilonatommy) [inline src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs:14]: utf-8 should be in encoded form in the code, they are more visible:
```suggestion
            yield return new object[] { "ar-SA", PlatformDetection.IsHybridGlobalizationOnBrowser ? "d\u200f/M\u200f/yyyy" : "d MMM\u060c yyyy g" }; // "d\u200f/M\u200f/yyyy g"
```

File: src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs [hunk 3]
```diff
@@ -33,7 +33,7 @@ public static IEnumerable<object[]> ShortDatePattern_Get_TestData_HybridGlobaliz
             yield return new object[] { "en-AG", "dd/MM/yyyy" };
             yield return new object[] { "en-AI", "dd/MM/yyyy" };
             yield return new object[] { "en-AS", "M/d/yyyy" };
-            yield return new object[] { "en-AT", "dd/MM/yyyy" };
+            yield return new object[] { "en-AT", PlatformDetection.IsHybridGlobalizationOnBrowser ? "dd/MM/yyyy" : "dd.MM.yyyy" }; // "dd/MM/yyyy"
```
Comment (ilonatommy) [inline src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs:36]: These differences were there even before we started normalizing? I don't see a place where Normalize/Fix methods would change `/` to `.`

Comment (matouskozak) [reply src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs:36]: Yes, this comes with "." from the Apple ICU.

**Review by ilonatommy — APPROVED** [review]

**Comment by matouskozak** [comment]
@tarekgh we *probably* hit the issue that different Apple ICU version give different outputs you mentioned here https://github.com/dotnet/runtime/issues/100240#issuecomment-2020774673. I plan to find a set of "stable" locals that I would use for testing. Do you think it would make sense to find a small set that could be used for ICU (non-hybrid) as well so that we would have coverage across Globalization backends or should I only focus on Apple hybrid in this PR? (Note: I planned to backport this PR to .NET 8)

File: src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs [hunk 4]
```diff
@@ -131,7 +137,6 @@ public static IEnumerable<object[]> ShortDatePattern_Get_TestData_HybridGlobaliz
             yield return new object[] { "en-ZA", "yyyy/MM/dd" };
             yield return new object[] { "en-ZM", "dd/MM/yyyy" };
             yield return new object[] { "en-ZW", "d/M/yyyy" };
-            yield return new object[] { "en-US", "M/d/yyyy" };
```
Comment (matouskozak) [inline src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortDatePattern.cs:134]: duplicate with line 125

**Comment by matouskozak** [comment]
> @matouskozak if `en-US` and `fr-FR` are not stable, I doubt any other cultures will be stable :-) these are the two culture that are sensitive to changes. I would suggest special case the hybrid.

Thanks. I tried invariant, `en-US` and `fr-FR` and they seem to be stable for now. For now, I only enabled it for Apple hybrid, we could later enable it for non-hybrid/ICU globalization as well (tested locally and it should work).

**Review by ilonatommy — APPROVED** [review]

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**